_posts/2025-11-24-asymmetric-distance-computation.md

```diff
@@ -0,0 +1,159 @@
+
+$$\textrm{total_bits} = \textrm{num_documents} \times \textrm{data_dimensionality}.$$
+
+To illustrate the magnitude of the savings, a dataset of 100 million different 768-dimensional vectors requires over 300 GB of system memory when using `float32` vectors, while its binary-quantized counterpart fits comfortably in under 10 GB.
+
+Unsurprisingly, these memory savings come with one major downside: recall degradation. The more you compress these vectors, the more difficult it becomes to estimate the distances between their ground-truth representations. This, in turn, makes it more difficult to identify the true k-nearest neighbors for a given query. While some of this downside is fundamentally unavoidable, various techniques have been discovered to help minimize the negative impact of data compression. In this blog post, we primarily focus on one such approach, *asymmetric distance computation (ADC)*, in which you can search through compressed indexes using uncompressed queries to attain 32-fold compression with improved recall compared to traditional binary quantization.
+
+## How binary quantization works
+
+With just one bit to encode each dimension, a binary-quantized vector can only assign one of two values to each dimension, typically interpreted as ±1. To quantize using binary quantization, the simplest method is to take the element-wise sign of the original vectors: negative values get mapped to -1 (encoded as `0` in binary), and positive values to +1 (encoded as `1`). 
+
+In practice, it is usually better to first *mean-center* the vectors, that is, shift the dataset toward the origin by subtracting the dataset (per-coordinate) mean. Quantization is then applied to these shifted vectors rather than the original. This helps preserve information about the original vectors, especially in datasets where some (or all!) coordinates may take the same sign. For example, every vector in the SIFT dataset has only non-negative values. Without mean-centering, all points would be binary-encoded to the all-ones vector. With mean-centering, the data would be represented using a wider variety of binary representations. 
+
+The shift-and-quantize process is outlined in the following figure.
+
+![Mean-Centering](/assets/media/blog-images/2025-11-24-asymmetric-distance-computation/Mean-Centering.png){:class="img-centered"}
+
+*Figure: To compute the binary quantization of a document corpus, we first mean-center the data so that the average document vector rests at the origin, and then convert each coordinate to a ±1 value depending on its sign.*
```

Comment:
```suggestion
*Figure: To compute the binary quantization of a document corpus, we first mean-center the data so that the average document vector rests at the origin and then convert each coordinate to a ±1 value, depending on its sign.*
```